nomad/job_endpoint.go

```diff
@@ -967,156 +967,188 @@ func (j *Job) BatchDeregister(args *structs.JobBatchDeregisterRequest, reply *st
+	if err != nil {
+		return err
+	}
+
+	// Find job
 	snap, err := j.srv.fsm.State().Snapshot()
 	if err != nil {
 		return err
 	}
+
 	ws := memdb.NewWatchSet()
 	job, err := snap.JobByID(ws, namespace, args.JobID)
 	if err != nil {
 		j.logger.Error("unable to lookup job", "error", err)
 		return err
 	}
+
 	if job == nil {
 		return structs.NewErrRPCCoded(404, fmt.Sprintf("job %q not found", args.JobID))
 	}
 
-	var found *structs.TaskGroup
+	// Find target group in job TaskGroups
+	groupName := args.Target[structs.ScalingTargetGroup]
+	var group *structs.TaskGroup
 	for _, tg := range job.TaskGroups {
-		if groupName == tg.Name {
-			found = tg
+		if tg.Name == groupName {
+			group = tg
 			break
 		}
 	}
-	if found == nil {
+
+	if group == nil {
 		return structs.NewErrRPCCoded(400,
 			fmt.Sprintf("task group %q specified for scaling does not exist in job", groupName))
 	}
 
-	now := time.Now().UnixNano()
-
-	// If the count is present, commit the job update via Raft
-	// for now, we'll do this even if count didn't change
-	prevCount := found.Count
-	if args.Count != nil {
-
-		// if there is a scaling policy, check that the new count is within bounds
-		if found.Scaling != nil {
-			if *args.Count < found.Scaling.Min {
-				return structs.NewErrRPCCoded(400,
-					fmt.Sprintf("group count was less than scaling policy minimum: %d < %d",
-						*args.Count, found.Scaling.Min))
-			}
-			if found.Scaling.Max < *args.Count {
-				return structs.NewErrRPCCoded(400,
-					fmt.Sprintf("group count was greater than scaling policy maximum: %d > %d",
-						*args.Count, found.Scaling.Max))
-			}
-		}
+	prevCount := int64(group.Count)
 
-		// Lookup the latest deployment, to see whether this scaling event should be blocked
-		d, err := snap.LatestDeploymentByJobID(ws, namespace, args.JobID)
+	// Handle non- count-based scaling event
+	if args.Count == nil {
+		_, eventIndex, err := j.srv.raftApply(
+			structs.ScalingEventRegisterRequestType,
+			&structs.ScalingEventRequest{
+				Namespace: job.Namespace,
+				JobID:     job.ID,
+				TaskGroup: groupName,
+				ScalingEvent: &structs.ScalingEvent{
+					Time:          time.Now().UnixNano(),
+					PreviousCount: prevCount,
+					Message:       args.Message,
+					Error:         args.Error,
+					Meta:          args.Meta,
+				},
+			},
+		)
 		if err != nil {
-			j.logger.Error("unable to lookup latest deployment", "error", err)
+			j.logger.Error("scaling event create failed", "error", err)
 			return err
 		}
-		// explicitly filter deployment by JobCreateIndex to be safe, because LatestDeploymentByJobID doesn't
-		if d != nil && d.JobCreateIndex == job.CreateIndex && d.Active() {
-			// attempt to register the scaling event
-			JobScalingBlockedByActiveDeployment := "job scaling blocked due to active deployment"
-			event := &structs.ScalingEventRequest{
+
+		reply.JobModifyIndex = job.ModifyIndex
+		reply.Index = eventIndex
+
+		j.srv.setQueryMeta(&reply.QueryMeta)
+
+		return nil
```

Comment:
non-error early returns for non-errors within conditionals have been a source of bugs.  Specially with large functions like this one, engineers add code to the end and miss the early return condition.  Is there a way to avoid the return without nested conditionals?

Reply:
I don't think there is a way. I'll remove the early return.